# Instructions

Tonight you're going to practice working in Pandas. You'll walk through instantiating a `DataFrame`, reading data into it, looking at and examining that data, and then playing with it. We'll be using the data in the `data` folder located within this directory (it's the same wine data that we looked at during lecture). Typically, we use IPython notebooks like this for a very specific set of things - presentations and EDA. Tonight, as we'll be playing around with `Pandas`, much of what we'll be doing is considered EDA. Therefore, by using a notebook, we'll get a tighter feedback loop with our work than we would trying to write a script. But, in general, **we do not use IPython notebooks for development**. 

Below, we've put a set of questions and then a cell for you to work on answers. However, feel free to add additional cells if you'd like. Often it will make sense to use more than one cell for your answers. 

# Assignment Questions 

### Part 1 - The Basics of DataFrames

Let's start off by following the general workflow that we use when moving data into a DataFrame: 

    * Importing Pandas
    * Reading data into the DataFrame
    * Getting a general sense of the data

Your tasks, right now:

1. Import pandas
2. Read the wine data into a DataFrame. 
3. Use the `attributes` and `methods` available on DataFrames to answer the following questions: 
    * How many rows and columns are in the DataFrame?
    * What data type is in each column?
    * Are all of the variables continuous, or are any categorical?
    * How many non-null values are in each column?
    * What are the min, mean, max, median for all numeric columns?

In [8]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('ggplot')


In [15]:
#reading a file and creating dataframe
df_red = pd.read_csv('data/winequality-red.csv', delimiter=';')
df_white = pd.read_csv('data/winequality-white.csv', delimiter=';')

In [20]:
df_red.shape , df_white.shape #Here I show how many (rows,colums)

((1599, 12), (4898, 12))

In [22]:
df_white.head

<bound method NDFrame.head of       fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36           20.70      0.045   
1               6.3             0.300         0.34            1.60      0.049   
2               8.1             0.280         0.40            6.90      0.050   
3               7.2             0.230         0.32            8.50      0.058   
4               7.2             0.230         0.32            8.50      0.058   
5               8.1             0.280         0.40            6.90      0.050   
6               6.2             0.320         0.16            7.00      0.045   
7               7.0             0.270         0.36           20.70      0.045   
8               6.3             0.300         0.34            1.60      0.049   
9               8.1             0.220         0.43            1.50      0.044   
10              8.1             0.270         0.41            1.45      0.033  

In [25]:
#showing the data type for each column. No nun values in any of the rows. The quaility type is categorical.o
df_white.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
fixed acidity           4898 non-null float64
volatile acidity        4898 non-null float64
citric acid             4898 non-null float64
residual sugar          4898 non-null float64
chlorides               4898 non-null float64
free sulfur dioxide     4898 non-null float64
total sulfur dioxide    4898 non-null float64
density                 4898 non-null float64
pH                      4898 non-null float64
sulphates               4898 non-null float64
alcohol                 4898 non-null float64
quality                 4898 non-null int64
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


In [26]:
df_white.describe() #This describes the dataset in terms of means and so on. 

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

### Part 2 - Practice with Grabbing Data

Let's now get some practice with grabbing certain parts of the data. If you'd like some extra practice, try answering each of the questions in more than one way (because remember, we can often grab our data in a couple of different ways). 

1. Grab the first 10 rows of the `chlorides` column. 
2. Grab the last 10 rows of the `chlorides` column. 
3. Grab indices 264-282 of the `chlorides` **and** `density` columns. 
4. Grab all rows where the `chlorides` value is less than 0.10. 
5. Now grab all the rows where the `chlorides` value is greater than the column's mean (try **not** to use a hard-coded value for the mean, but instead a method).
6. Grab all those rows where the `pH` is greater than 3.0 and less than 3.5. 
7. Further filter the results from 6 to grab only those rows that have a `residual sugar` less than 2.0. 

In [67]:
#grabing the first 10 rows of the chlorides column

#but first changin the names of the colums and replacing the spaces with '-'
cols = df.columns.tolist()
cols = [col.replace(' ', '_') for col in cols]
df_red.columns = cols

#now doing the grabing the first ten rows of the chlorides.

#df_red.loc[:9, 'chlorides']



In [68]:
df_red.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [45]:
df_red['chlorides'].tail(10)

1589    0.073
1590    0.077
1591    0.089
1592    0.076
1593    0.068
1594    0.090
1595    0.062
1596    0.076
1597    0.075
1598    0.067
Name: chlorides, dtype: float64

In [46]:
#Grabing certain rows in the chlordes and density colums

df_red.loc[264:282, ['chlorides' , 'density']]

chlorides  density
264      0.064  0.99990
265      0.071  0.99680
266      0.096  1.00025
267      0.078  0.99730
268      0.077  0.99870
269      0.104  0.99960
270      0.087  0.99650
271      0.104  0.99960
272      0.071  0.99935
273      0.076  0.99735
274      0.088  0.99915
275      0.087  0.99650
276      0.077  0.99870
277      0.104  0.99960
278      0.073  0.99760
279      0.087  0.99910
280      0.071  0.99860
281      0.358  0.99720
282      0.067  0.99710

In [48]:
#Grabing all the rows where chlorides are less than 0.10
df_red[df_red['chlorides'] < 0.10]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
5               7.4             0.660         0.00             1.8      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
8               7.8             0.580         0.02             2.0      0.073   
9               7.5             0.500         0.36             6.1      0.071   
10              6.7             0.580         0.08             1.8      0.097   
11              7.5             0.500         0.36             6.1      0.071   
12              5.6             0.615         0.00             1.6      0.089   
16              8.5             0.280         0.56             1.8      0.092   
18              7.4             0.590         0.08             4.4      0.086   
20              8.9             0.220         0.48             1.8      0.077   
21              7.6             0.390         0.31             2.3      0.082   
23              8.5             0.490         0.11             2.3      0.084   
24              6.9             0.400         0.14             2.4      0.085   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
28              7.1             0.710         0.00             1.9      0.080   
29              7.8             0.645         0.00             2.0      0.082   
30              6.7             0.675         0.07             2.4      0.089   
32              8.3             0.655         0.12             2.3      0.083   
33              6.9             0.605         0.12            10.7      0.073   
35              7.8             0.645         0.00             5.5      0.086   
36              7.8             0.600         0.14             2.4      0.086   
37              8.1             0.380         0.28             2.1      0.066   
39              7.3             0.450         0.36             5.9      0.074   
...             ...               ...          ...             ...        ...   
1567            7.2             0.695         0.13             2.0      0.076   
1568            7.0             0.560         0.13             1.6      0.077   
1569            6.2             0.510         0.14             1.9      0.056   
1571            6.4             0.380         0.14             2.2      0.038   
1572            7.3             0.690         0.32             2.2      0.069   
1573            6.0             0.580         0.20             2.4      0.075   
1574            5.6             0.310         0.78            13.9      0.074   
1575            7.5             0.520         0.40             2.2      0.060   
1576            8.0             0.300         0.63             1.6      0.081   
1577            6.2             0.700         0.15             5.1      0.076   
1579            6.2             0.560         0.09             1.7      0.053   
1580            7.4             0.350         0.33             2.4      0.068   
1581            6.2             0.560         0.09             1.7      0.053   
1582            6.1             0.715         0.10             2.6      0.053   
1583            6.2             0.460         0.29             2.1      0.074   
1584            6.7             0.320         0.44             2.4      0.061   
1585            7.2             0.390         0.44             2.6      0.066   
1586            7.5             0.310

In [56]:
#Now grab all the rows where the chlorides value is greater than the column's mean 
#(try not to use a hard-coded value for the mean, but instead a method)

chlorides_mean = df_red['chlorides'].mean()
df_red[df_red['chlorides'] > chlorides_mean]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
10              6.7             0.580         0.08             1.8      0.097   
12              5.6             0.615         0.00             1.6      0.089   
13              7.8             0.610         0.29             1.6      0.114   
14              8.9             0.620         0.18             3.8      0.176   
15              8.9             0.620         0.19             3.9      0.170   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
19              7.9             0.320         0.51             1.8      0.341   
22              7.9             0.430         0.21             1.6      0.106   
27              7.9             0.430         0.21             1.6      0.106   
30              6.7             0.675         0.07             2.4      0.089   
31              6.9             0.685         0.00             2.5      0.105   
34              5.2             0.320         0.25             1.8      0.103   
38              5.7             1.130         0.09             1.5      0.172   
41              8.8             0.610         0.30             2.8      0.088   
42              7.5             0.490         0.20             2.6      0.332   
46              7.7             0.935         0.43             2.2      0.114   
47              8.7             0.290         0.52             1.6      0.113   
54              7.6             0.510         0.15             2.8      0.110   
57              7.5             0.630         0.12             5.1      0.111   
61              7.7             0.690         0.49             1.8      0.115   
71              7.7             0.670         0.23             2.1      0.088   
74              9.7             0.320         0.54             2.5      0.094   
75              8.8             0.410         0.64             2.2      0.093   
76              8.8             0.410         0.64             2.2      0.093   
77              6.8             0.785         0.00             2.4      0.104   
81              7.8             0.430         0.70             1.9      0.464   
83              7.3             0.670         0.26             1.8      0.401   
...             ...               ...          ...             ...        ...   
1391            8.0             0.640         0.22             2.4      0.094   
1395            8.6             0.685         0.10             1.6      0.092   
1396            8.7             0.675         0.10             1.6      0.090   
1413            9.9             0.570         0.25             2.0      0.104   
1423            6.4             0.530         0.09             3.9      0.123   
1434           10.2             0.540         0.37            15.4      0.214   
1435           10.2             0.540         0.37            15.4      0.214   
1436           10.0             0.380         0.38             1.6      0.169   
1441            7.4             0.785         0.19             5.2      0.094   
1445            7.4             0.785         0.19             5.2      0.094   
1451            7.8             0.320         0.44             2.7      0.104   
1474            9.9             0.500         0.50            13.8      0.205   
1476            9.9             0.500         0.50            13.8      0.205   
1479            8.2             0.280         0.60             3.0      0.104   
1481            8.2             0.280         0.60             3.0      0.104   
1482            7.2             0.580         0.54             2.1      0.114   
1493            7.7             0.540         0.26             1.9      0.089   
1496            7.7             0.540

In [58]:
df_red.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [60]:
#Grab all those rows where the pH is greater than 3.0 and less than 3.5.
df_red[(df_red['pH'] > 3.0) & (df_red['pH'] < 3.5 )]


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
8               7.8             0.580         0.02             2.0      0.073   
9               7.5             0.500         0.36             6.1      0.071   
10              6.7             0.580         0.08             1.8      0.097   
11              7.5             0.500         0.36             6.1      0.071   
13              7.8             0.610         0.29             1.6      0.114   
14              8.9             0.620         0.18             3.8      0.176   
15              8.9             0.620         0.19             3.9      0.170   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
18              7.4             0.590         0.08             4.4      0.086   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
22              7.9             0.430         0.21             1.6      0.106   
23              8.5             0.490         0.11             2.3      0.084   
24              6.9             0.400         0.14             2.4      0.085   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
29              7.8             0.645         0.00             2.0      0.082   
30              6.7             0.675         0.07             2.4      0.089   
31              6.9             0.685         0.00             2.5      0.105   
32              8.3             0.655         0.12             2.3      0.083   
33              6.9             0.605         0.12            10.7      0.073   
34              5.2             0.320         0.25             1.8      0.103   
...             ...               ...          ...             ...        ...   
1560            7.8             0.600         0.26             2.0      0.080   
1561            7.8             0.600         0.26             2.0      0.080   
1562            7.2             0.695         0.13             2.0      0.076   
1563            7.2             0.695         0.13             2.0      0.076   
1564            7.2             0.695         0.13             2.0      0.076   
1565            6.7             0.670         0.02             1.9      0.061   
1566            6.7             0.160         0.64             2.1      0.059   
1567            7.2             0.695         0.13             2.0      0.076   
1568            7.0             0.560         0.13             1.6      0.077   
1569            6.2             0.510         0.14             1.9      0.056   
1570            6.4             0.360         0.53             2.2      0.230   
1571            6.4             0.380         0.14             2.2      0.038   
1572            7.3             0.690         0.32             2.2      0.069   
1574            5.6             0.310         0.78            13.9      0.074   
1575            7.5             0.520         0.40             2.2      0.060   
1576            8.0             0.300         0.63             1.6      0.081   
1578            6.8             0.670         0.15             1.8      0.118   
1580            7.4             0.350

In [63]:
#Further filter the results from 6 to grab only those rows that have a residual sugar less than 2.0.
df_red[(df_red['pH'] > 3.0) & (df_red['pH'] < 3.5 ) & (df_red['residual sugar'] < 2.0 )]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
3              11.2             0.280         0.56             1.9      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
10              6.7             0.580         0.08             1.8      0.097   
13              7.8             0.610         0.29             1.6      0.114   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
22              7.9             0.430         0.21             1.6      0.106   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
34              5.2             0.320         0.25             1.8      0.103   
44              6.8             0.670         0.02             1.8      0.050   
47              8.7             0.290         0.52             1.6      0.113   
48              6.4             0.400         0.23             1.6      0.066   
49              5.6             0.310         0.37             1.4      0.074   
50              8.8             0.660         0.26             1.7      0.074   
61              7.7             0.690         0.49             1.8      0.115   
62              7.5             0.520         0.16             1.9      0.085   
66              7.5             0.520         0.11             1.5      0.079   
67              6.6             0.705         0.07             1.6      0.076   
69              8.0             0.705         0.05             1.9      0.074   
70              7.7             0.630         0.08             1.9      0.076   
72              7.7             0.690         0.22             1.9      0.084   
79              8.3             0.625         0.20             1.5      0.080   
80              6.2             0.450         0.20             1.6      0.069   
81              7.8             0.430         0.70             1.9      0.464   
...             ...               ...          ...             ...        ...   
1457            7.6             0.490         0.33             1.9      0.074   
1458            8.4             0.290         0.40             1.7      0.067   
1459            7.9             0.200         0.35             1.7      0.054   
1464            6.8             0.590         0.10             1.7      0.063   
1465            6.8             0.590         0.10             1.7      0.063   
1473            6.1             0.600         0.08             1.8      0.071   
1483            8.1             0.330         0.44             1.5      0.042   
1490            7.1             0.220         0.49             1.8      0.039   
1493            7.7             0.540         0.26             1.9      0.089   
1494            6.4             0.310         0.09             1.4      0.066   
1495            7.0             0.430         0.02             1.9      0.080   
1496            7.7             0.540         0.26             1.9      0.089   
1500            7.5             0.725         0.04             1.5      0.076   
1509            7.9             0.180         0.40             1.8      0.062   
1513            6.4             0.560         0.15             1.8      0.078   
1521            6.9             0.480         0.20             1.9      0.082   
1527            7.1             0.530         0.07             1.7      0.071   
1533            8.0             0.390

### Part 3 - More Practice

Let's move on to some more complicated things. Use your knowledge of `groupby`s, `sorting`, and the other things that you learned in lecture to answer the following. 

1. Get the average amount of `chlorides` for each `quality` value. 
2. For observations with a `pH` greater than 3.0 and less than 4.0, find the average `alcohol` value by `pH`. 
3. For observations with an `alcohol` value between 9.25 and 9.5, find the highest amount of `residual sugar`. 
4. Create a new column, called `total_acidity`, that is the sum of `fixed acidity` and `volatile acidity`. 
5. Find the average `total_acidity` for each of the `quality` values. 
6. Find the top 5 `density` values. 
7. Find the 10 lowest `sulphates` values. 

### Part 4 - Practice with Plotting 

1. Plot the average amount of `chlorides` for each `quality` value (1 from Part 3). 
2. Plot the `alcohol` values against `pH` values. Does there appear to be any relationship between the two?
3. Plot `total_acidity` values against `pH` values. Does there appear to be any relationship between the two?
4. Plot a histogram of the `quality` values. Are they evenly distributed within the data set?
5. Plot a boxplot to look at the distribution of `citric acid`. 

### Part 5 - Putting it All Together 

Now that you've worked on all the basics with one data set, it's time to do it with a second! This time, though, you'll go through the process of downloading the data set yourself. You'll also go through the process of learning to ask questions of the data (i.e. you won't be given any questions). We'll point you to a number of different data sets, and let you go at it. In reality, this is often how data science works. There isn't a clear-cut set of instructions on what to do - you kind of just dive into the data and see what you find!

Your goal by the end of `Part 5` is to be able to tell a story with your data. Whether that means you query it and find something interesting, examine a number of different columns and their values, or plot a couple of different columns, it doesn't matter. You should aim to find at least one piece of interesting information in your data (and ideally even more than one). Then, tell your peers and the instructors what you've found! 

Potential data sources: 

1. [Forest-fires](http://archive.ics.uci.edu/ml/datasets/Forest+Fires)
2. [Iris](http://archive.ics.uci.edu/ml/datasets/Iris)
3. [Another wine data set](http://archive.ics.uci.edu/ml/datasets/Wine)
4. [Abalone](http://archive.ics.uci.edu/ml/datasets/Abalone)
5. [Adult Income data set](http://archive.ics.uci.edu/ml/datasets/Adult)

The links above are all to the home pages of these data sources. At the top of these pages, you will find a link the the `Data Folder` where you can actually find the data. The majority of these data sets don't come in `.csv` format. While one of the datasets is available in `.csv` format, we encourage you to pick whatever data set you find most interesting (regardless of the format), and challenge yourself to read the necessary documentation and go through the process of figuring out how to get the data from the web and into a `DataFrame` (the instructors will also be around to help).